clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3430,23 +3430,98 @@ static void handleSYCLIntelSchedulerTargetFmaxMhzAttr(Sema &S, Decl *D,
-  if (D->isInvalidDecl())
-    return;
-
-  Expr *E = A.getArgAsExpr(0);
+void Sema::AddSYCLIntelMaxGlobalWorkDimAttr(Decl *D,
+                                            const AttributeCommonInfo &CI,
+                                            Expr *E) {
+  if (!E->isValueDependent()) {
+    // Validate that we have an integer constant expression and then store the
+    // converted constant expression into the semantic attribute so that we
+    // don't have to evaluate it again later.
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return;
+    E = Res.get();
 
-  if (!checkWorkGroupSizeValues(S, D, A)) {
-    D->setInvalidDecl();
-    return;
+    // This attribute requires a non-negative value.
+    if (ArgVal < 0) {
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*non-negative*/ 1;
+      return;
+    }
+    // This attribute must be in the range [0, 3].
+    if (ArgVal > 3) {
+      Diag(E->getBeginLoc(), diag::err_attribute_argument_out_of_range)
+          << CI << 0 << 3 << E->getSourceRange();
+      return;
+    }
+    // Check to see if there's a duplicate attribute with different values
+    // already applied to the declaration.
+    if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
+      // If the other attribute argument is instantiation dependent, we won't
+      // have converted it to a constant expression yet and thus we test
+      // whether this is a null pointer.
+      const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getValue());
+      if (DeclExpr && ArgVal != DeclExpr->getResultAsAPSInt()) {
+        Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+        Diag(DeclAttr->getLoc(), diag::note_previous_attribute);
+        return;
+      }
+    }
+    // Checks correctness of mutual usage of different work_group_size
+    // attributes: reqd_work_group_size, max_work_group_size and
+    // max_global_work_dim.
+    // In case the value of 'max_global_work_dim' attribute equals to 0
+    // we shall ensure that if max_work_group_size and reqd_work_group_size
+    // attributes exist, they hold equal values (1, 1, 1).
+    auto Check = [this, &CI](const auto *A) {
+      Optional<llvm::APSInt> XDimVal = A->getXDimVal(Context);
+      Optional<llvm::APSInt> YDimVal = A->getYDimVal(Context);
+      Optional<llvm::APSInt> ZDimVal = A->getZDimVal(Context);
+      if (*XDimVal != 1 || *YDimVal != 1 || *ZDimVal != 1) {
+        Diag(A->getLocation(), diag::err_sycl_x_y_z_arguments_must_be_one)
+            << A << CI;
+        return false;
+      }
+      return true;
+    };
+    if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+      if ((ArgVal == 0) && !Check(DeclAttr))
+        return;
+    }
+    if (const auto *DeclAttr = D->getAttr<ReqdWorkGroupSizeAttr>()) {
+      if ((ArgVal == 0) && !Check(DeclAttr))
```

Comment:
I am checking value of max_global_work_dim here instead of lambda function because it will not be captured there. instead of adding duplicate codes, i have decided to check here.

Reply:
This looks sensible to me, though I'd drop the parens around `ArgVal == 0` as they're no longer necessary (in both places).